src/lib389/cli/dscontainer

```diff
@@ -352,6 +355,19 @@ binddn = cn=Directory Manager
 
     log.info("389-ds-container started.")
 
+    # If oneshot parameter is passed, run an optional script and
+    # shutdown the server.
+    if args.oneshot != False:
+        if args.oneshot != None:
+            if os.path.exists(args.oneshot):
+                log.info("Running oneshot script...")
+                oneshot_p = subprocess.Popen(args.oneshot, stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE, env=os.environ.copy())
+                oneshot_p.wait()
+            else:
+                log.info("Oneshot script is not found!")
+        kill_ds()
```

Comment:
Why immediately shutdown rather than leaving the instance running?

Reply:
Because I want to run `dscontainer` in oneshot mode at the container *build* time, rather than as entry point. So that the resulting container will have a preconfigured instance. 
Here's an example: https://github.com/vashirov/ds-container/blob/dscontainer_oneshot/fedora-minimal/Containerfile.f33#L21

Reply:
The problem is that, any content you change in the volume (or /data which becomes the volume) will not always be overlaid into the volume that is consumed by the user. So you can't configure at build time, you can only rely on configuration to be applied at first run. That's why it would be better to have a stateful expression because this could be run multiple times and you can't guarantee that the changes will be present after a build. 

So while it may work for you to have this run at build time, it won't work for *other* people. Which is a bit why I'm worried about making a feature like this which could have a lot of sharp edges ...

Reply:
> The problem is that, any content you change in the volume (or /data which becomes the volume) will not always be overlaid into the volume that is consumed by the user. 

Can you please elaborate and provide an example? Is this still a problem in a throwaway-ldap-container-for-testing use case where there are no external volumes?

> So you can't configure at build time, you can only rely on configuration to be applied at first run. 

Are you talking about `dscontainer` in particular or about containers in general?

> That's why it would be better to have a stateful expression because this could be run multiple times and you can't guarantee that the changes will be present after a build.

Could you please help me understand what change do you want me to do? Or provide a concrete practical example why the current approach can go wrong?

> So while it may work for you to have this run at build time, it won't work for _other_ people. Which is a bit why I'm worried about making a feature like this which could have a lot of sharp edges ...

From what I saw, other people are doing the same thing with our DS containers, but with extra steps (`dscreate` and fake `systemctl` scripts). So why not make their life easier?

Reply:
> Because I want to run `dscontainer` in oneshot mode at the container _build_ time, rather than as entry point. So that the resulting container will have a preconfigured instance.
> Here's an example: https://github.com/vashirov/ds-container/blob/dscontainer_oneshot/fedora-minimal/Containerfile.f33#L21

To clarify: I run `dscontainer` in oneshot mode during the build, and then it is set as a CMD in a run mode.

Reply:
> > The problem is that, any content you change in the volume (or /data which becomes the volume) will not always be overlaid into the volume that is consumed by the user.
> 
> Can you please elaborate and provide an example? Is this still a problem in a throwaway-ldap-container-for-testing use case where there are no external volumes?

It's not a problem for testing or throw-away. It is a problem for release builds though, or people building and layering things together. 

> 
> > So you can't configure at build time, you can only rely on configuration to be applied at first run.
> 
> Are you talking about `dscontainer` in particular or about containers in general?

Containers in general. 

> 
> > That's why it would be better to have a stateful expression because this could be run multiple times and you can't guarantee that the changes will be present after a build.
> 
> Could you please help me understand what change do you want me to do? Or provide a concrete practical example why the current approach can go wrong?

My concern is that while the one shot is a quick solution to your problem that you want to solve, it will be used by other people in the wild. Containers should be driven by stateful configuration management where you express a desired state, rather than "please run these things to get to this state". I think that the oneshot mode has a lot of ways it "could go wrong" in production to users of the containers 

> 
> > So while it may work for you to have this run at build time, it won't work for _other_ people. Which is a bit why I'm worried about making a feature like this which could have a lot of sharp edges ...
> 
> From what I saw, other people are doing the same thing with our DS containers, but with extra steps (`dscreate` and fake `systemctl` scripts). So why not make their life easier?

I would assume they are doing this because we don't have a good way to provide stateful configuration into the container yet, so that's why I think that working on a way to do this that solves both cases would be really good - but I accept it would certainly be "more work" than what you have proposed here.

Reply:
Okay, I see your point. I agree that stateful configuration would be a way to go. And yes, it would be a non-trivial amount of work to get it done. I will open a new ticket to track it. I will also open a new ticket to fix a syntax warning and usage information that were fixed in this PR.

Thanks!